students/jared_mulholland/lesson_6/test_mail_room_class_2.py

```diff
@@ -0,0 +1,156 @@
+    dg.add_donor = chris
+
+    assert "Chris Cornell" in dg.donor_list 
+
+def test_create_report():
+    """test that report is actually created"""
+    jared = Donor("Jared Mulholland", 10000) 
+    chris = Donor("Chris Cornell", 50000)
+
+    dg = DonorGroup(jared)
+    dg.add_donor = chris
+
+    x = dg.create_report
+    assert x == "report created"
+
+def test_send_letters():
+    """test that the letters are sent by checking number of files in folder after function is run"""
+    jared = Donor("Jared Mulholland", 10000) 
+    chris = Donor("Chris Cornell", 50000)
+    ben = Donor("Ben Shepard", 4000)
+    kim = Donor("Kim Thayil", 3400)
+
+    dg = DonorGroup(jared)
+    dg.add_donor = chris
+    dg.add_donor = kim
+    dg.add_donor = ben
+
+    donor_count = len(dg.donor_list)
+
+    letter_dir = "C:\\Users\\Jared\\Documents\\IntroToPython\\Self_Paced-Online\\students\\jared_mulholland\\lesson_9\\donor_letters"
+    dg.send_letters = letter_dir
+    assert donor_count == len(os.listdir(letter_dir))
+
+    filelist = [f for f in os.listdir(letter_dir)]
+    for f in filelist:
+        os.remove(os.path.join(letter_dir, f))
+
+
+"""test functions in mail room:
+    1. send_thankyou()
+    2. create_report(): create report tested in DonorGroup class tests
+    3. send_letters(): send letters testing in DonorGroup class tests """
+
+#def test_send_letters():
+#    input = lambda input: ['Eddie Vedder', 10000]
+#    assert 'Eddie' in send_thankyou()
+
+
+#Lesson 10 additions
+"""
+1.
+Add a new feature to Mailroom using map so that each donation on record can be doubled, tripled or indeed multiplied by any 
+arbitrary factor based on the whims of philanthropists who would like to support our cause.
+
+This will require a new function (or method in your donor database class) called challenge(factor) that takes a multiplier (factor), 
+and multiplies all the donations of all the donors by the factor. The function returns a NEW donor database, with the new data.
+
+2. 
+Add a new feature to Mailroom using filter so that donations either above or below a specified dollar amount are included in the map operations of #1 above.
+You can do this by adding min_donation and max_donation optional keyword parameters to your challenge function. You’ll want to filter the donations before 
+passing them to map.
+"""
+
+
+def test_factor_donation():
+    """tests challenge function to see if factor works correctly"""
+
+    billy = Donor("Billy Corgan", [10,20])
+    james = Donor("James Iha", [30,100])
+    mark = Donor("Mark Lanegan", [50, 150, 40])
+
+    dg_test = DonorGroup(billy)
+    dg_test.add_donor = james
+    dg_test.add_donor = mark
+    factor = 2
+    factored_amount = 800
+    assert factored_amount ==  dg_test.projection(factor)
+
+
+def test_factored_donation_min_max():
+
+    factor = 2
+    min_don = 11
+    max_don = 99
+    factored_amount = 280
+    assert factored_amount == dg_test.projection(factor, min_don, max_don)
```

Comment:
dg_test object should be created in the test as it could have weird side-effects